src/sparseml/transformers/utils/sparse_model.py

```diff
@@ -250,58 +258,43 @@ def text_generation_from_pretrained(
         """
-        SparseAutoModel._check_tf(model_name_or_path)
-        if not kwargs:
-            kwargs = {}
-        kwargs["from_tf"] = False
-        delayed = False
-        if "state_dict" not in kwargs:
-            kwargs["state_dict"], delayed = SparseAutoModel._loadable_state_dict(
-                model_name_or_path
-            )
+        if config.model_type == "opt":
+            # TODO: Talk to Alex whether this pathway needs to be maintained
+            def skip(*args, **kwargs):
+                pass
 
-        kwargs["config"].is_decoder = True
-        kwargs["config"].use_past = False
+            torch.nn.init.kaiming_uniform_ = skip
+            torch.nn.init.uniform_ = skip
+            torch.nn.init.normal_ = skip
 
-        model = AutoModelForCausalLM.from_pretrained(
-            model_name_or_path,
-            **kwargs,
-        )
-        SparseAutoModel.log_model_load(model, model_name_or_path, model_type, delayed)
+            model = OPTForCausalLM.from_pretrained(
+                model_name_or_path,
+                torch_dtype=torch_dtype,
+                config=config,
+                trust_remote_code=trust_remote_code,
+                **kwargs,
+            )
+        else:
+            model = AutoModelForCausalLM.from_pretrained(
+                model_name_or_path,
+                torch_dtype=torch_dtype,
+                config=config,
+                trust_remote_code=trust_remote_code,
+                **kwargs,
+            )
 
-        return model
+        # TODO: Do we need to call eval() here? Why?
+        model.eval()
+        model.seqlen = sequence_length
 
-    @staticmethod
-    def text_generation_from_pretrained_distil(
-        model_name_or_path: str,
-        teacher_name_or_path: Optional[str],
-        model_kwargs: Dict[str, Any],
-        teacher_kwargs: Dict[str, Any],
-    ) -> Tuple[Module, Optional[Module]]:
-        """
-        :param model_name_or_path: the name of or path to the model to load
-        :param teacher_name_or_path: the name of or path to the teacher to load,
-            None or one of ['self', 'disable'] will not create a teacher and
-            instead return the value passed in
-        :param model_kwargs: the keyword args to pass into the AutoModel for model
-        :param teacher_kwargs: the keyword args to pass into the AutoModel for teacher
-        :return: a tuple containing the model and distillation teacher (optional)
-            for text generation
-        """
-        model = SparseAutoModel.text_generation_from_pretrained(
-            model_name_or_path,
-            model_type="student" if teacher_name_or_path else "model",
-            **model_kwargs,
-        )
-        teacher = (
-            SparseAutoModel.text_generation_from_pretrained(
-                teacher_name_or_path, model_type="teacher", **teacher_kwargs
+        if recipe:
+            apply_structure_to_transformers(
```

Comment:
@bfineran this function both applies the structure and loads weight, confirmed by @Satrat